CIP (#5)
will-break-it/wallet-architecture: docs/CIP/CIP-XXXX/README.md

- the current tip/ block height for validity intervals of transactions as well as sync progress

#### How do we serve this data?

Through a push-based API, wallets are provided with a continuous stream of **relevant** on-chain events, enabling them to derive their current wallet state efficiently. This API supports various blockchain models, including UTXO-based chains and account-based chains. Additionally, clients receive secondary data relevant to the blockchain network they are subscribed to, ensuring comprehensive and up-to-date information.

#### Protocol

We use a bidirectional protocol and encode all messages as JSON objects, while transactions contained in messages are preserved in the chain's native encoding standard (cbor). Messages adhere to a specific structure, ensuring consistency and facilitating efficient communication within our event-driven protocol. As we strive to support multiple blockchains, a server-sent message will always reference the originating chain it relates to.

### Message Types & Structure

We distinguish between client-sent and server-sent messages.

#### Client

Client-sent messages are **typed** messages, such as:

> ```json

Reviewer: ```suggestion
```json5
```
Does this work?

#### Account Authentication

Clients connect and authenticate via subscribe messages by providing:

- a digital signature
- a public key whose Blake2b-256 hash corresponds to one of their provided credentials: payment and stake key hash(es) or script hash (es) in case of shared wallets

This ensures that only authorized clients can access their account data.

The `signature` field is generated by signing with the private payment/ stake key a SHA256 HMAC hashed string. The prehash string is constructed by concatenating the subscribed blockchain + timestamp.

The server verifies the signature, timestamp and whether the Blake2b-256 hash of public key matches one of the credentials before serving any data to the client. If any part of the `subscribe` message is invalid, the server sends an error and closes the connection.

#### Transaction Querying

The credentials also enable efficient querying of client-relevant data:

1. **Indexing Strategy**

   We index transactions by both payment and stake credential. This dual-indexing approach allows for comprehensive transaction tracking.

2. **BIP32 Wallet Query Process**

>    The server queries our index using the corresponding key hashes per client to serve only relevant transactions to scubribed clients.

Reviewer: ```suggestion
   The server queries our index using the corresponding key hashes per client to serve only relevant transactions to subscribed clients.
```

### Versioning

We propose to version message partials individually. In our protocol, the client sends its list of supported versions as part of the initial [`subscribe`](./messages/client/subscribe.md) message. Upon receiving this information, the server selects the highest mutually supported version and associates it with the client.
This method requires the server to implement version-based message construction, allowing it to tailor its responses to each client's supported version.

### Encoding

Our proposed communication protocol **adopts blockchain-native encoding standards** to ensure compatibility and efficiency. This approach leverages existing conventions familiar to clients while minimizing conversion overhead on the server-side. (see [CPS - Fragile API](../../CPS/CPS-XXXX/README.md#fragile-api))

#### Address Encoding

Address encoding varies depending on the blockchain network. In Cardano the protocol servers bech32 encoded addresses.

> ### Extensions

Reviewer: We need to define how extensions will be supported under this standard. Best to take inspiration from https://github.com/cardano-foundation/CIPs/pull/462.

---
CIP: 1

> Title: Event-driven, wallet-opimised API

Reviewer: ```suggestion
Title: Event-driven, wallet-optimized API
```

#### Server

Instead of typed messages, server-sent messages vary in structure depending on the underlying blockchain. Our design aims to be chain-agnostic and serve data from different blockchains in future. Therefore, we rely on **untyped** messages that are enriched and aggregated into compound messages. We call each top level key a server message partial.

This leads to a structure that has many top level keys of which, not every key may be present for every message, for example:

```json
{
  "transactions": [
    /* ... */
  ],
  "resolvedInputs": [
    /* ... */
  ],
  "rewards": {
    /* ... */
  },
  "tip": {
    /* ... */
  }
  // ...
}
```

> [!IMPORTANT]
>
> New message partials can be added at any time, which allows for flexible extension of the API.
> Clients are expected to check for message key presence and ignore those they do not support.

#### Errors

Any cases of server-side failure trigger an error message that is append as separate `error` message partial:

```json
{
  "error": {
    "type": "error type",
    "message": "error message"
  }
  /* ... */
}
```

### Authentication

> A newly connected client authenticates as part of the [`subscribe`](./messages/client/subscribe.md) message that is sent to the server to synchronize a specific wallet account up to the chain's current tip. This message must include a **signature** that verifies the ownership of provided extened public key (xpub key).

Reviewer: ```suggestion
A newly connected client authenticates as part of the [`subscribe`](./messages/client/subscribe.md) message that is sent to the server to synchronize a specific wallet account up to the chain's current tip. This message must include a **signature** that verifies the ownership of provided extended public key (xpub key).
```

Extended public keys (xpub keys) serve a dual purpose in our design: **authentication** and **efficient transaction querying**. Here's how it works:

#### Account Authentication

Clients connect and authenticate via subscribe messages by providing:

- a digital signature
- their extended account public key (xpub key)

This ensures that only authorized clients can access their account data.

> The `signature` field is generated by signing with the [extended account private key](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) a SHA256 HMAC hashed string. The prehash string is constructed by concatenating the subscribed blockchain(s) + timestamp.

Reviewer 2: I think we agreed on supporting one blockchain per subscription. To prevent confusion:
```suggestion
The `signature` field is generated by signing with the [extended account private key](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) a SHA256 HMAC hashed string. The prehash string is constructed by concatenating the subscribed blockchain + timestamp.
```

> The server verifies the signature and timestamp before serving any data to the client. If any part of the `subscribe` message is invalid, the server sends an error and closes the connection.

Reviewer 2: Should we mention a timestamp window since we will use `timestamp` for signature verification and try to guess it on the server? Should the timestamp be in a 60ms window, meaning the server tries 60 times before failing?

3. **BIP32 Wallet Query Process**

   - The server derives address credentials from the client's xpub key
   - It then queries our index using the corresponding key hashes
   - This process continues until reaching a predefined [gap limit](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery)

> [!NOTE]
> **Gap Limit:**
> We stop querying after encountering a specified number of consecutive unused addresses (the gap limit). This prevents unnecessary searches while ensuring we capture all relevant transactions.

### Ordering of Events

Our API needs a reliable measurement of time to construct a stream of ordered, client-relevant, on-chain events that connected clients can easily consume. Time is crucial for clients for various reasons, including:

- monitoring the progress of synchronization up to the chain's tip
- defining transaction validity intervals

>
> - supports operations that need to occur within specific timeframes
> - helps manage network congestion by allowing expired transactions to be discarded
> - protects against replay attacks by ensuring transactions can't be resubmitted after expiration

> We borrow the term `point` from the Cardano blockchain to introduce a sever-sent message partial. For instance, if a new transaction is deemed relevant for a connected client, the server enriches the message by adding the current "`point` in on-chain time" of the respective chain. For Cardano, this typically looks as follows:

Reviewer 2: typo in `sever-sent`:
```suggestion
We borrow the term `point` from the Cardano blockchain to introduce a server-sent message partial. For instance, if a new transaction is deemed relevant for a connected client, the server enriches the message by adding the current "`point` in on-chain time" of the respective chain. For Cardano, this typically looks as follows:
```

---
CIP: 1
Title: Event-driven, wallet-opimised API
Status: Proposed
Category: Wallets
Authors:
  - William Wolff <william.wolff@iohk.io>

>   - Sefa Irken <sefa.irken@iohk.io>

Reviewer: ```suggestion
  - Hızır Sefa İrken <sefa.irken@iohk.io>
```

  - Rhys Bartels-Waller <rhys.bartelswaller@iohk.io>
  - Martynas Kazlauskas <martynas.kaslauskas@iohk.io>
  - Daniele Ricci <daniele.ricci@iohk.io>
Implementors: N/A
Solution-To: CPS-????
Discussions:
Created: 2024-07-10
License: CC-BY-4.0
---

## Abstract

> We propose an event-driven, push-based API optimized for wallet applications, leveraging the principle that a wallet's state should be a pure function of the blockchain state. This approach eliminates polling strategies and aggregating views, enabling real-time synchronization with the blockchain as new blocks are appended. Our solution addresses key challenges in current wallet implementations, including rollback handling, address discovery and multi-account support.

Reviewer: ```suggestion
We propose a multi-chain, event-driven, push-based API optimized for wallet applications, leveraging the principle that a wallet's state should be a pure function of the blockchain state. This approach eliminates polling strategies and aggregating views, enabling real-time synchronization with the blockchain as new blocks are appended. Our solution addresses key challenges in current wallet implementations, including rollback handling, address discovery and multi-account support.
```

## Motivation: why is this CIP necessary?

The CIP provides a partial solution to the problems described in [CPS-????](../CPS/CPS-XXXX/README.md). In particular, we strive to define a standardized approach serving data to edge clients in an unopinionated, close to blockchain native format using a push-based protocol.

## Specification

### Introduction

#### What does a wallet need to construct transactions?

> - past transactions (receiving & spending) to derive balance and UTxO set

Reviewer 3: ```suggestion
- past transactions (receiving & spending) to derive balance, UTxO set, delegation and governance state
```

> - rewards (staking, voting)

Reviewer 3: Technically in order to build transactions you only need the current balance
```suggestion
- reward account balance (staking, voting)
```

- network, era or epoch specific data, like:
  - protocol parameters (tx fees, plutus cost models, ...)
  - era summaries (slot length)

>   - genesis (security param)

Reviewer 3: Era summaries is an abstraction over some genesis parameters. I don't think you need historical/past slot length when constructing new transactions.
```suggestion
  - genesis (security param, slot length)
```

Author: Fair point!

- the current tip/ block height for validity intervals of transactions as well as sync progress

#### How do we serve this data?

Through a push-based API, wallets are provided with a continuous stream of **relevant** on-chain events, enabling them to derive their current wallet state efficiently. This API supports various blockchain models, including UTXO-based chains and account-based chains. Additionally, clients receive secondary data relevant to the blockchain network they are subscribed to, ensuring comprehensive and up-to-date information.

#### Protocol

> We use a bidirectional protocol (websockets) and encode all messages as JSON objects, while transactions contained in messages are preserved in the chain's native encoding standard (cbor). Messages adhere to a specific structure, ensuring consistency and facilitating efficient communication within our event-driven protocol. As we strive to support multiple blockchains, a server-sent message will always reference the originating chain it relates to.

Reviewer: ```suggestion
We use a bidirectional protocol and encode all messages as JSON objects, while transactions contained in messages are preserved in the chain's native encoding standard (cbor). Messages adhere to a specific structure, ensuring consistency and facilitating efficient communication within our event-driven protocol. As we strive to support multiple blockchains, a server-sent message will always reference the originating chain it relates to.
```

I feel it's better to not specify the exact protocol here and just describe it from a high level, so that we could see implementations in other protocols, such as gRPC.

Reviewer 3: Had almost the exact same comment in my review draft :sweat_smile:

### Message Types & Structure

We distinguish between client-sent and server-sent messages.

#### Client

Client-sent messages are **typed** messages, such as:

```json
{

>   "type": "<unique message type>>"

Reviewer 4: ```suggestion
  "type": "<unique message type>"
```
... just a typo

  // ...
}
```

This allows the server to quickly identify the client's intent. More details on what typed messages exist will be covered in [Client Typed Messages](#client-typed-messages) section.

#### Server

Instead of typed messages, server-sent messages vary in structure depending on the underlying blockchain. Our design aims to be chain-agnostic and serve data from different blockchains in future. Therefore, we rely on **untyped** messages that are enriched and aggregated into compound messages. We call each top level key a server message partial.

This leads to a structure that has many top level keys of which, not every key may be present for every message, for example:

```json
{
  "transactions": [
    /* ... */
  ],
  "resolvedInputs": [
    /* ... */
  ],
  "rewards": {
    /* ... */
  },
  "tip": {
    /* ... */
  }
  // ...
}
```

> [!IMPORTANT]
>
> New message partials can be added at any time, which allows for flexible extension of the API.

> > Clients are expected to iterate over message keys and skip those they do not support.

Reviewer 3: Seems like clients [specify extensions on subscription](https://github.com/will-break-it/wallet-architecture/tree/cip/docs/CIP/CIP-XXXX#extensions). Also, not sure if 'iterate' is the right word. It's a JSON object (not an array), so most of the time clients would check the specific keys they are interested in.
```suggestion
```

Author: The client doesn't know what keys a message will include, so the implementation will likely use something like `Object.keys(message).forEach(partialKey => ...)` which is an iteration over message keys.

Reviewer 3: I would do `message[key]` instead. At least in js you can easily check if it's defined or not.

#### Errors

Any cases of server-side failure trigger an error message that is append as separate `error` message partial:

```json
{
  "error": {
    "type": "error type",
    "message": "error message"
  }
  /* ... */
}
```

### Authentication

A newly connected client authenticates as part of the [`subscribe`](./messages/client/subscribe.md) message that is sent to the server to synchronize a specific wallet account up to the chain's current tip. This message must include a **signature** that verifies the ownership of provided extened public key (xpub key).

Extended public keys (xpub keys) serve a dual purpose in our design: **authentication** and **efficient transaction querying**. Here's how it works:

> #### Account Authentication

Reviewer 3: How do you authenticate for tracking script credentials?

Clients connect and authenticate via subscribe messages by providing:

- a digital signature
- their extended account public key (xpub key)

This ensures that only authorized clients can access their account data.

The `signature` field is generated by signing with the [extended account private key](https://input-output-hk.github.io/adrestia/static/Ed25519_BIP.pdf) a SHA256 HMAC hashed string. The prehash string is constructed by concatenating the subscribed blockchain(s) + timestamp.

The server verifies the signature and timestamp before serving any data to the client. If any part of the `subscribe` message is invalid, the server sends an error and closes the connection.

#### Transaction Querying

The xpub key also enables efficient querying of client-relevant data:

1. **Address Derivation**

>    Using the client's xpub key, our server can deterministically derive a series of public addresses associated with the client's wallet.

Reviewer 3: We'd probably be doing 'credential discovery' instead of traditional address discovery
```suggestion
   Using the client's xpub key, our server can deterministically derive a series of public address credentials associated with the client's wallet.
```

2. **Indexing Strategy**

>    We index transactions by both payment key hash and stake key hash. This dual-indexing approach allows for comprehensive transaction tracking.

Reviewer 3: For script wallet support
```suggestion
   We index transactions by both payment and stake credential. This dual-indexing approach allows for comprehensive transaction tracking.
```

3. **Query Process**

>    - The server derives addresses from the client's xpub key

Reviewer 3: ```suggestion
3. **BIP32 Wallet Query Process**

   - The server derives address credentials from the client's xpub key
```

   - It then queries our index using the corresponding key hashes

>    - This process continues until reaching a predefined [gap limit](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#account-discovery)

Reviewer 4: If we index transaction by keys (rather than by addresses) here we can get a couple of enhancements:

1. there's no need for the server to derive addresses: clients subscribes to keys, transactions are indexed by keys...
2. but moreover we can completely overcome the discovery process which leads us to completely ignore the _gap limit issue_; as soon as new address appears on the network, all its transactions are already indexed by keys and the queries get them even if far more than twice the gap limit.

> [!NOTE]
> **Gap Limit:**

> > We stop querying after encountering a specified number of consecutive unused addresses (the gap limit). This prevents unnecessary searches while ensuring we capture all relevant transactions.

Reviewer 4: This is what Lace does today i.e. the discovery process. This is no longer needed if transactions are indexed by keys.

### Ordering of Events

Our API needs a reliable measurement of time to construct a stream of ordered, client-relevant, on-chain events that connected clients can easily consume. Time is crucial for clients for various reasons, including:

- monitoring the progress of synchronization up to the chain's tip
- defining transaction validity intervals

> [!NOTE]

> > Transaction validity intervals allow clients to submit transactions with a specified lifetime, which serves several important purposes:

Reviewer 3: Imho this is just noise in context of this CIP
```suggestion
```

>
> - supports operations that need to occur within specific timeframes
> - helps manage network congestion by allowing expired transactions to be discarded
> - protects against replay attacks by ensuring transactions can't be resubmitted after expiration

We borrow the term `point` from the Cardano blockchain to introduce a sever-sent message partial. For instance, if a new transaction is deemed relevant for a connected client, the server enriches the message by adding the current "`point` in on-chain time" of the respective chain. For Cardano, this typically looks as follows:

#### Cardano Point

```json
{
  "point": {
    "slot": 127838345,
    "hash": "9f06ab6ecce25041b3a55db6c1abe225a65d46f7ff9237a8287a78925b86d10e"
  }
  /* ... */
}

> ```

Reviewer 4: Are we sure we don't want to include the block height?
All the tools show it: wallets, chain explorers...

Reviewer: The _point_  serves a particular purpose, which is to capture the current "time". Anything outside of this would be considered unnecessary and we would be bleeding into this concept being more of a _block header_

> [!NOTE]

> > Clients may safely assume that any serve-sent message, aside from error messages contain `point` partial.

Reviewer 4: ```suggestion
> Clients may safely assume that any server-sent message, aside from error messages contain `point` partial.
```
typo

### Synchronization

> We refer to synchronization as the process that a wallet undertakes to catch up to the current chain's tip in order to derive its latest state and proceed to transact. The afore-described `point` represents the main anchor for wallets, to be able to:

Reviewer: ```suggestion
We refer to synchronization as the process that a wallet undertakes to catch up to the current chain's tip in order to derive its latest state and proceed to transact. The aforedescribed `point` represents the main anchor for wallets, to be able to:
```

> [!NOTE]
> More details are described in the [`subscribe`](./messages/client/subscribe.md) message.

#### Rollback Handling

In a [CAP](https://en.wikipedia.org/wiki/CAP_theorem) system like Cardano, which balances global consistency with availability, there is a challenge in dealing with rollback events that impact transaction finality. Regardless of whether one uses their own full node or a server provider for transaction submission - if such events are not properly handled with the help of monitoring the rolling window of [k blocks](https://plutus-apps.readthedocs.io/en/latest/plutus/explanations/rollback.html) and respective resubmission, there is no guarantee that the transaction is finalized, eg. becomes part of the immutable pefix of the Cardano blockchain.

> Since our wallet-optimized protocol defines that server-sent messages include the `point` message partial, the client can check if its last `point` is greater than any subsequent message `point` from the server, without the need to introduce a `rollback` specific message partial. This assumes that the client keeps a local copy of the mutuable suffix of the blockchain defined by the security parameter.

Reviewer: ```suggestion
Since our wallet-optimized protocol defines that server-sent messages include the `point` message partial, the client can check if its last `point` is greater than any subsequent message `point` from the server, without the need to introduce a `rollback` specific message partial. This assumes that the client keeps a local copy of the volatile segment of the blockchain defined by the security parameter.
```

#### Connection Management

Clients are expected to reconnect on error, when providing invalid `point`s for which a server cannot find an intersection.
Clients may send multiple [`subscribe`](./messages/client/subscribe.md) messages for different accounts, but subsequently take responsibility to coordinate server-sent events by time to their possibly differently synchronized accounts. When receiving a message that references a `point` that the first account has already seen, the client only applies it for its second account for example.

### Versioning

> In our protocol, the client sends its list of supported versions as part of the initial [`subscribe`](./messages/client/subscribe.md) message. Upon receiving this information, the server selects the highest mutually supported version and associates it with the client.

Reviewer: It would be best to define what is versioned before talking about how the negotiation takes place.

This method requires the server to implement version-based message construction, allowing it to tailor its responses to each client's supported version.

### Encoding

> Our proposed communication protocol **adopts blockchain-native encoding standards** to ensure compatibility and efficiency. This approach leverages existing conventions familiar to clients while minimizing conversion overhead on the server-side.

Reviewer: Best to use this opportunity to address the problem statement's section on [Fragile API](https://github.com/will-break-it/wallet-architecture/blob/main/docs/CPS/CPS-XXXX/README.md#fragile-api)

> #### Transaction Hash Encoding

Reviewer 3: Where is transaction hash used in the API?
```suggestion
```

Reviewer 3: I suppose in the resolved inputs partial :thinking: although that would be best to document in the extension

Transaction hashes are typically encoded using **hexadecimal representation**.

#### Address Encoding

Address encoding varies depending on the blockchain network. In Cardano the protocol servers bech32 encoded addresses.

### Extensions

Clients can turn on certain `extensions` by setting a respective flags in their subscription [`config](./messages/client/subscribe.md#subscribe-example) which tells the server to for example:

- to enrich transaction output asset values by their respective off-chain metadata such as decimal offsets
- to resolve transaction inputs (transaction output references) as transaction outputs

## Rationale: how does this CIP achieve its goals?

A primary consideration for the protocol design was to minimize message round trips between client and server as well as the amount of work to be done by the server for each client. Thus, the API's authentication does not require an initial message exchange but rather defines a standard verifable method for the server to either drop a client early and proceed to publish its relevant data.

### Client Typed Messages

> - [`heartbeat`](./messages/client/heartbeat.md): Connection management/ keep alive

Reviewer 3: This is a transport-specific message. Are we sure it's even needed with WS? I'd expect the connection to be maintained by either the protocol or lower level libraries.
```suggestion
```

Author: WebSocket connections can drop due to network issues or infrastructure timeouts. Proxies and firewalls may terminate idle WebSocket connections too.

Reviewer 3: Infrastructure can be configured. Anyway, even if this is actually required for WS, it's transport-specific. So based on https://github.com/will-break-it/wallet-architecture/pull/5#discussion_r1678761352 this is probably not the right place to define this message? :thinking:

Reviewer: You could make note that since the API supports extensions,  transport-specific messages can be implemented such as this.

- [`subscribe`](./messages/client/subscribe.md): Define topics of interest & authentication

> - [`unsubscribe`](./messages/client/unsubscribe.md): Graceful disconnect from server

Reviewer 4: I see this as an useless message which only effect is to increase the complexity without advantages.

- I don't see a real use case where a wallet doesn't want to stop _following_ transactions for its addresses.
- Most likely clients may choose to simply disconnect without caring about this: being no advantages for them in implementing this, there is an high probability many companies will prefer to save money to implement it.
- The server needs to unsubscribe the client on disconnect anyway:
  - to protect from unexpected close,
  - to not waste resources for lazily wrote clients which simply disconnect.

I'd simply remove it.

Reviewer 2: @iccicci If you have ten accounts in your wallet and want to remove one of them, you might want to be able to unsubscribe.

Reviewer 4: Yes @ginnun , in theory yes... but which is the use case?

Since others can still send transactions to addresses bound to that account, what does it means "I want to remove an account"?

To me sounds as "if somebody still send me coins on one of these addresses, I'm no longer interested: it's fine if I do not track them".

When I want to stop using an account, I'll stop sharing addresses bound to it with other to ask for payments, but I'm still interested (much interested) in transactions which give me something.

What am I missing?
Thank you

Reviewer 2: You are still the account owner, but you are no longer interested in monitoring it on a given subscription. One person can have multiple installations of the wallet, e.g., mobile and desktop, simultaneously. 

A wallet will probably allow you to remove a wallet address via the UI (Lace does), why should server still send events related to it? 

- After some time, you may want to decrease the number of subscriptions in your mobile wallet while keeping everything active on the desktop. You may want to hide it for some reason.
- You may want to drop an unused subscription because you want to.

Reviewer 4: Makes sense...
Probably we can just mark it as optional for the clients.

### Server Message Partials

- [`genesis`](./messages/server/genesis.md): If applicable for the subscribed blockchain, it serves static data used as part of the network bootstrap
- [`transaction`](./messages/server/transaction.md): A new transaction
- [`tip`](./messages/server/tip.md): A new block was appended
- [`protocol parameters`](./messages/server/protocol-parameters.md): Epoch based message for updated protocol parameters

> - [`era summary`](./messages/server/era-summary.md): Era based message for updated slot length

Reviewer 3: Since the credentials are no longer supplied by the client, server should send the credentials it discovered. Otherwise client won't know it's own addresses.
```suggestion
- [`era summary`](./messages/server/era-summary.md): Era based message for updated slot length
- credentials: credentials discovered by the server
```

Reviewer 4: Sorry @mkazlauskas , I do not understand this.

> Since the credentials are no longer supplied by the client

Didn't we say that clients subscribe through credentials?

I thought the flow should be something similar to:

1. the client asks the server to follow all transactions relevant for all its credentials;
  - `subscribe` with a set of credentials (so credentials are provided by the client);
2. the server answers with the list of all addresses bound to the credentials set;
  - having transactions indexed by credentials, the server can find all of them using the indexes rather than the standard discovery process;
  - rather than adding `credentials: credentials discovered by the server` probably we can add `addresses: addresses discovered by the server`.

Reviewer 3: > Didn't we say that clients subscribe through credentials?

That was the original idea, but then we added subscribing through xpub which is not a credential, but is a base key that can be used to derive the key hash credentials
